File: Remove_Nth_Node_From_End_of_List.cpp
/**
 * Definition of ListNode
 * class ListNode {
 * public:
 *     int val;
 *     ListNode *next;
 *     ListNode(int val) {
 *         this->val = val;
 *         this->next = NULL;
 *     }
 * }
 */
class Solution {
public:
    /**
     * @param head: The first node of linked list.
     * @param n: An integer.
     * @return: The head of linked list.
     */
    ListNode *removeNthFromEnd(ListNode *head, int n) {
        // write your code here
        //maybe the n is larger than the length of head
        auto dummy = new ListNode(-1);
        dummy->next = head;
        auto pre = dummy;
        int length = 0;
        while (pre->next != nullptr){
            ++length;
            pre = pre->next;
        }
        n = n % length;
        n = n == 0 ? length : n;
        pre = dummy;
        while (n > 0){
            --n;
            pre = pre->next;
        }
        auto next = dummy;
        while (pre->next != nullptr){
            pre = pre->next;
            next = next->next;
        }
        //remove node
        auto node_to_delete = next->next;
        next->next = node_to_delete->next;
        //delete the removed node
        delete node_to_delete;
        return dummy->next;
    }
};

File: Remove Nth Node From End of List.py
"""
Definition of ListNode
class ListNode(object):

    def __init__(self, val, next=None):
        self.val = val
        self.next = next
"""
class Solution:
    """
    @param head: The first node of linked list.
    @param n: An integer.
    @return: The head of linked list.
    """
    def removeNthFromEnd(self, head, n):
        # write your code here
        dummy = ListNode(-1)
        dummy.next = head
        fast = dummy
        for __ in range(n):
            fast = fast.next
        slow = dummy
        while fast.next:
            fast, slow = fast.next, slow.next
        
        slow.next = slow.next.next
        return dummy.next


## new solution and solid codes
    def removeNthFromEnd(self, head, n):
        # write your code here
        if not head:
            return None
        if n <= 0:
            return head
        
        # check if n <= length
        length = 0
        cur = head
        while cur:
            length += 1
            cur = cur.next
        
        if length < n:
            return head
        
        if n == length:
            return head.next
            
        n = length - n + 1 - 1
        cur = head
        # while n > 1:
        for __ in range(n - 1):
            cur = cur.next
        cur.next = cur.next.next
        return head